src/features/congregation/upcoming_events/upcoming_events_list/useUpcomingEventsList.tsx

```diff
@@ -1,49 +1,113 @@
 
 const useUpcomingEventsList = ({ data }: UpcomingEventsListProps) => {
-  const [eventsSortedByYear, setEventsSortedByYear] = useState([[]]);
+  const stickyYearRefs = useRef([]);
+  const [stuckYearIndexes, setStuckYearIndexes] = useState<Set<number>>(
+    new Set()
+  );
 
-  const sortEventsByYear = (events: UpcomingEventType[]) => {
-    if (events.length === 0) {
-      return [[]];
-    }
+  const { isAdmin } = useCurrentUser();
 
-    const tmpStack: Record<number, UpcomingEventType[]> = {};
+  const offsetTopMap = useRef<Map<number, number>>(new Map());
 
-    events.forEach((event) => {
-      if (!event.event_data.date) {
-        return;
-      }
-      const year = new Date(event.event_data.date).getFullYear();
+  const isEventExpired = useCallback(
+    (event: UpcomingEventType, isAdmin: boolean): boolean => {
+      const currentDate = new Date();
 
-      if (!tmpStack[year]) {
-        tmpStack[year] = [];
-      }
+      return event.event_data.event_dates.every((event_date) => {
+        const endDate = new Date(event_date.end);
+        const diffInTime = currentDate.getTime() - endDate.getTime();
+        const diffInDays = diffInTime / (1000 * 3600 * 24);
 
-      tmpStack[year].push(event);
-    });
+        return isAdmin ? diffInDays > 5 : diffInDays > 1;
+      });
+    },
+    []
+  );
 
-    const keys = Object.keys(tmpStack);
+  const sortEventsByYear = useCallback(
+    (events: UpcomingEventType[]) => {
+      const yearMap = new Map<number, UpcomingEventType[]>();
 
-    if (keys.length === 0) {
-      return [[]];
-    } else if (keys.length === 1) {
-      return keys.map((year) => tmpStack[Number(year)]);
-    } else {
-      return keys
-        .toSorted((a, b) => Number(a) - Number(b))
-        .map((year) => tmpStack[Number(year)]);
-    }
-  };
+      for (const event of events) {
+        const dateStr = event.event_data.event_dates?.[0]?.start;
+        if (!dateStr || event._deleted || isEventExpired(event, isAdmin))
+          continue;
+
+        const year = new Date(dateStr).getFullYear();
+        if (!yearMap.has(year)) {
+          yearMap.set(year, []);
+        }
+        yearMap.get(year)!.push(event);
+      }
+
+      const sortedYears = Array.from(yearMap.keys()).sort((a, b) => a - b);
+      return sortedYears.map((year) => yearMap.get(year)!);
+    },
+    [isAdmin, isEventExpired]
+  );
+
+  const eventsSortedByYear = useMemo(() => {
+    return sortEventsByYear(data);
+  }, [data, sortEventsByYear]);
 
   useEffect(() => {
-    setEventsSortedByYear(sortEventsByYear(data));
-  }, [data]);
+    const handleScroll = () => {
+      const newStuckSet = new Set<number>();
+
+      stickyYearRefs.current.forEach((element, index) => {
+        if (!element || !element.parentElement) {
+          return;
+        }
+
+        if (!offsetTopMap.current.has(index)) {
+          const offset = element.offsetTop - 56;
+          if (offset !== 50) {
```

Comment:
will this still work depending on the screen size?  and can we use constants for these variables?